Risk engine initialisation, update from legacy risk engine workflow and status change

nkhristinin opened the pull request:
## Risk engine initialisation, update from legacy risk engine workflow and status change


https://github.com/elastic/kibana/assets/7609147/dfb75d4a-f447-4346-9760-d0e9685cce39


Green areas it is what was implemented
<img width="1449" alt="Screenshot 2023-08-01 at 15 07 01" src="https://github.com/elastic/kibana/assets/7609147/4d87887f-1163-45eb-a4e9-a77a685f6565">


This pr has:
- Upgrade workflow. If the user has a risk host or user transforms, we will show the panel with a call to action for the upgrade.
- Introduce new Saved object to save the configuration of risk engine
- API which is described bellow

It required experiment enabled - **riskScoringRoutesEnabled**
## New API

### /engine/status

#### GET
Get the status of the Risk Engine

##### Description:
Returns the status of both the legacy transform-based risk engine, as well as the new risk engine

##### Responses

```json
{
  "legacy_risk_engine_status": "NOT_INSTALLED" , "ENABLED"
  ,
  "risk_engine_status": "NOT_INSTALLED" , "ENABLED" , "DISABLED"
}
```

### /engine/init

#### POST
Initialize the Risk Engine

##### Description:
Initializes the Risk Engine by creating the necessary indices and mappings, removing old transforms, creating saved object configuration

##### Responses

```json
{
  "result": {
    "risk_engine_enabled": true,
    "risk_engine_resources_installed": true,
    "risk_engine_configuration_created": true,
    "legacy_risk_engine_disabled": true,
    "errors": [
      "string"
    ]
  }
}
```

### /engine/enable

#### POST
Enable the Risk Engine
##### Description:
Change saved object configuration and in the future here we will start task

### /engine/disable

#### POST
Disable the Risk Engine
Change saved object configuration and in the future here we will stop task

nkhristinin committed:
Risk engine API start

nkhristinin committed:
Add saved objects and status

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/check_mappings_update --fix'

nkhristinin committed:
Error handling

nkhristinin committed:
Return who last updated

nkhristinin committed:
Add risk_score_update_panel

nkhristinin committed:
Risk update panel

nkhristinin committed:
Add risk score update panels

nkhristinin committed:
Delete old transforms

nkhristinin committed:
Add mapping for SO

nkhristinin committed:
fix name

nkhristinin committed:
Fix types

nkhristinin committed:
type

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

nkhristinin committed:
add tests

nkhristinin committed:
Wrong rebase

nkhristinin committed:
clean

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/risk_engine_data_client.ts:
```diff
@@ -37,10 +62,58 @@ interface RiskEngineDataClientOpts {
 }
 
+interface Configuration {
+  enable: boolean;
```
```suggestion
  enabled: boolean;
```

"Enabled" is an adjective describing the risk engine; "enable" is a verb that doesn't read as well 👍

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/risk_engine_data_client.ts:
```diff
@@ -57,10 +130,209 @@ export class RiskEngineDataClient {
+      const configuration = savedObjectConfiguration?.attributes;
+
+      if (configuration) {
+        return configuration as Configuration;
```
I think you can type  `getConfigurationSavedObject`'s return and remove the need for this cast.

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/schema/risk_score_apis.yml:
```diff
@@ -44,6 +44,56 @@ paths:
+  /engine/status:
+    get:
+      summary: Get the status of the Risk Engine
+      description: Returns the status of the legacy transforms risk engine as well as the new risk engine
```
```suggestion
      description: Returns the status of both the legacy transform-based risk engine, as well as the new risk engine
```

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/schema/risk_score_apis.yml:
```diff
@@ -44,6 +44,56 @@ paths:
+                $ref: '#/components/schemas/RiskEngineStatusResponse'
+                
+  /engine/init:
+    posrt:
```
```suggestion
    post:
```

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/routes/risk_engine_init_route.ts:
```diff
@@ -0,0 +1,69 @@
+      const siemResponse = buildSiemResponse(response);
+      const securitySolution = await context.securitySolution;
+      const soClient = (await context.core).savedObjects.client;
+      const riskEgineClient = securitySolution.getRiskEngineDataClient();
```
```suggestion
      const riskEngineClient = securitySolution.getRiskEngineDataClient();
```

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/schema/risk_score_apis.yml:
```diff
@@ -153,7 +203,46 @@ components:
+        result:
+          type: object
+          properties:
+            riskEngineEnabled:
```
Can we snake_case these? The camelCased fields are inconsistent with the other APIs.

nkhristinin replied:
I can:) just feel a bit wrong that we manipulate camelCase on the server side and clients side, but for API layer should make it different

rylnd replied:
Agreed, but that's always been the way (unless we forget 😅).

nkhristinin commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into enable-risk-score

nkhristinin commented on x-pack/plugins/security_solution/public/entity_analytics/api/api.ts:
```diff
@@ -27,3 +39,44 @@ export const fetchRiskScorePreview = async ({
+/**
+ * Fetches risks engine status
+ */
+export const fetchRiskEngineStatus = async ({
```
My original idea was to introduce the new public folder, for all things related to EA (watchlist, etc)

If you think we should rename/move things please let me know

nkhristinin commented on x-pack/plugins/security_solution/public/entity_analytics/api/hooks/use_disable_risk_engine_mutation.ts:
```diff
@@ -0,0 +1,35 @@
+  EnableDisableRiskEngineResponse,
+} from '../../../../server/lib/risk_engine/types';
+
+export const DISABLE_RISK_ENGINE_MUTATION_KEY = ['POST', 'DISABLE_RISK_ENGINE'];
```
At least for useQuery, we can use those keys for invalidate query, but for mutation, I am not sure, what can be a case, but I saw that at least for detection engine we also export those keys for mutations

nkhristinin commented on x-pack/plugins/security_solution/public/entity_analytics/translations.ts:
```diff
@@ -130,3 +130,117 @@ export const PREVIEW_QUERY_ERROR_TITLE = i18n.translate(
+export const ERROR_PANEL_MESSAGE = i18n.translate(
+  'xpack.securitySolution.riskScore.errorPanel.message',
+  {
+    defaultMessage: 'Something wen’t wrong. Try again later.',
```
@craig-abbott What do you think?

nkhristinin commented on x-pack/plugins/security_solution/server/lib/risk_engine/risk_engine_data_client.ts:
```diff
@@ -57,10 +130,209 @@ export class RiskEngineDataClient {
+      savedObjectConfiguration.id,
+      {
+        ...attributes,
+        last_updated_by: user?.username ?? '',
```
I think saved_object itself has this date for the last update.

Probably when we have a solution for a chronological history of actions, then we should also save what type of actions was made. But for now, probably it should be enough.

SourinPaul replied:
++ @rylnd It would be ideal to display enablement/ disablement activity separately from subsequent risk configuration changes. 

Typically, the enablement is a Detection Engineer (DE) persona and Risk Configuration can (Analyst and DE). Although we don't have the RBAC granularity for enforcing, displaying this explicitly in the audit log will reduce user confusion.

Does this answer the question?

nkhristinin commented on x-pack/plugins/security_solution/server/lib/risk_engine/risk_engine_data_client.ts:
```diff
@@ -37,10 +62,58 @@ interface RiskEngineDataClientOpts {
   private writerCache: Map<string, Writer> = new Map();
   constructor(private readonly options: RiskEngineDataClientOpts) {}
 
+  public async init({ namespace, savedObjectsClient, user }: InitOpts) {
```
Can you elaborate please, why we need to do that and what exact logic should we move?

I understand that file is a bit big, but looks like logic is more or less related to this class.

The only things I thought made sense to move, were its exact operation with saved objects/transforms.

My concern, with extracting it to a lot of files, is that it will be harder to understand.

rylnd replied:
In my experience, a pure function with dependencies explicitly passed in is always going to be easier to reason about than a class method leveraging a bunch of internal state/methods. 

The dependency tree for `init` is quite complicated, and this class is getting quite large. I observed that `init` receives a `savedObjectsClient`, which seems to be the center of much of this new code, and so I thought that it might be the correct place to abstract some logic into a dedicated function.

Writing this class in terms of declarative, dependency-injected functions all but guarantees those logical units can be reused later, and forces you to consider the dependencies/interaction between them by not allowing you to hide that complexity through shared state and embedded method calls. The class, then, only serves as the coordination point and central calling point of that code.

As an example: `disableLegacyTransforms` looks like:

* `disableLegacyTransforms()` : `namespace`
  * `getLegacyStatus()`: `namespace`
    * `getLegacyTransforms()`: `namespace`
  * `esClient`
  * `getLegacyTransforms()`: `namespace`
    * `esClient`

While it would be more work than leaving it as is (of course, that's always true), this could all be moved into a `risk_engine_legacy_transforms` file if `disableLegacyTransforms` and `getLegacyTransforms` are modified to accept an `esClient` instead of pulling it from an instance property.

rylnd replied:
> The only things I thought made sense to move, were its exact operation with saved objects/transforms.

I think I agree with that, but is that not the majority of code you've added here, besides the public methods themselves? I feel like we're saying the same thing, but perhaps I've misunderstood.

rylnd replied:
> My concern, with extracting it to a lot of files, is that it will be harder to understand.

I think the entire point of my suggestion is to identify logical groupings of code, and abstract/group them as such. Right now, the class does several things, and logic is spread amongst a multitude of private methods. At the very least, splitting that out into `risk_engine_legacy_transforms` and `risk_engine_saved_objects` will ease the mental burden of parsing through all that in a single file 👍

nkhristinin committed:
Typos and PR fixes

nkhristinin committed:
add api tets

nkhristinin committed:
fix unit tests

nkhristinin committed:
Try to enable feature in cypress tests

nkhristinin committed:
udpate mappings

nkhristinin commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into enable-risk-score

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

nkhristinin committed:
Refactoring

nkhristinin commented on x-pack/plugins/security_solution/public/entity_analytics/components/risk_score_enable_section.tsx:
```diff
@@ -40,40 +57,186 @@ const docsLinks = [
+            </EuiFlexGroup>
+          </EuiModalHeader>
+        )}
+        {!initRiskEngineMutation.isLoading && (
```
React can render `0` for example if you have something like `{arr.length && (<div>content</div>))`

In our case it shouldn't be a problem but anyway I change to ternary operation, I think it make more sense here

nkhristinin committed:
Fix jest tests

kibanamachine committed:
Merge branch 'main' into enable-risk-score

nkhristinin committed:
Try to fix types

jbudz reviewed (approved):
`packages/kbn-test` LGTM

nkhristinin commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into enable-risk-score

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/utils/risk_engine_transforms.ts:
```diff
@@ -0,0 +1,73 @@
+  esClient: ElasticsearchClient;
+}) => {
+  const getTransformStatsRequests: Array<Promise<TransformGetTransformResponse>> = [];
+  [RiskScoreEntity.host, RiskScoreEntity.user].forEach((entity) => {
```
Since `RiskScoreEntity` is a string enum, you should be able to do: 
```suggestion
  Object.values(RiskScoreEntity).forEach((entity) => {
```

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/utils/risk_engine_transforms.ts:
```diff
@@ -0,0 +1,73 @@
+
+  const transforms = fulfuletGetTransformStats.reduce((acc, val) => {
+    if (val.transforms) {
+      return [...acc, ...val.transforms];
```
I'm sure you know this, but I'll say it for posterity:

It's typically bad practice to instantiate a new object on each loop of a reduce, if the real goal is to append/augment an existing object. It's absolutely fine in this case with just four transforms, but I know we had a [PR](https://github.com/elastic/kibana/pull/157471) a while ago that updated all of these similarly, and this one is particularly easy:

```suggestion
      return acc.push(...val.transforms);
```

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/utils/risk_engine_transforms.ts:
```diff
@@ -0,0 +1,73 @@
+  const result = await Promise.allSettled(getTransformStatsRequests);
+
+  const fulfuletGetTransformStats = result
+    .filter((r) => r.status === 'fulfilled')
```
You can make this a type predicate to eliminate the `as` cast below:
```suggestion
    .filter((r): r is PromiseFulfilledResult<TransformGetTransformResponse> => r.status === 'fulfilled')
```

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/utils/risk_engine_transforms.ts:
```diff
@@ -0,0 +1,73 @@
+
+  const fulfuletGetTransformStats = result
+    .filter((r) => r.status === 'fulfilled')
+    .map((r) => (r as PromiseFulfilledResult<TransformGetTransformResponse>).value);
```
More broadly: this whole "filter, map, reduce" pipeline is a little meandering; why not just something like:

```ts
results
  .filter((result) => result.status === 'fulfilled' && result.value.transforms)
  .flatMap((result) => result.value.transforms);
```

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/utils/risk_engine_transforms.ts:
```diff
@@ -0,0 +1,73 @@
+  return transforms;
+};
+
+export const removeLegacyTransofrms = async ({
```
```suggestion
export const removeLegacyTransforms = async ({
```

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/utils/risk_engine_transforms.ts:
```diff
@@ -0,0 +1,73 @@
+}: {
+  namespace: string;
+  esClient: ElasticsearchClient;
+}) => {
```
```suggestion
}): Promise<void> => {
```

I appreciate the explicit return type, I don't always want to read/wait for the compiler 😉

rylnd replied:
Also: do we not care about the results of this call? I suppose we verify by retrieving their presence/absence afterward?

nkhristinin replied:
Yes, there is some API, which failed, because transforms do not exist for example, so it's a bit easier to check later

rylnd reviewed (commented):
I think the only thing that _needs_ to be addressed here is adding `dynamic: false` on the SO mapping; the others are _shoulds_: 

* logic of parsing the transform response (esp. the reduce/spread combo)
* various typos
* better (or explicit) typings

The nits, of course, you can take or leave. 😄 

I'll give a quick approval once the above are addressed (and I expect you'll have some time to do this while you track down the other code owners for review 😉 ).

Great work, thank you @nkhristinin!

nkhristinin commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into enable-risk-score

nkhristinin committed:
Change enable risk button

nkhristinin committed:
Add cy tests

nkhristinin committed:
PR fixes

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/check_mappings_update --fix'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

rylnd reviewed (approved):
Thanks for addressing my changes. Let me know if you'd like review on anything else, but this looks good to go on my end. 👍

SourinPaul commented on x-pack/plugins/security_solution/public/entity_analytics/translations.ts:
```diff
@@ -130,3 +130,117 @@ export const PREVIEW_QUERY_ERROR_TITLE = i18n.translate(
+export const UPDATE_RISK_ENGINE_MODAL_EXISTING_USER_HOST_2 = i18n.translate(
+  'xpack.securitySolution.riskScore.updateRiskEngineModal.existingUserHost_2',
+  {
+    defaultMessage: ', they are no longer required.',
```
@rylnd I see that the [UX copy](https://www.figma.com/file/jwzKiq2IyxapRlAzmz1kyJ/8.9-%7C-6546-%7C-Migrate-to-new-scoring-engine-%7C-Entity-Analytics?type=design&node-id=143-84665&mode=design&t=h99GdItjePQ6bRFZ-0) specifies the reason as: 

`Existing user and host risk score transforms will be deleted, they are not required for the new entity risk engine.`

IMO, it reads adequately. cc: @craig-abbott in case I missed anything.

craig-abbott replied:
@rylnd @SourinPaul I'm happy to go with the more concise "as they are no longer required."

In terms of explaining why they are no longer required, I'd tried to allude to that by saying they're not needed for the new risk engine, but it doesn't really go into detail, hence I guess you wanting to make it more concise haha. 

There is a link to "learn more about updating the new risk engine" where I assumed we could provide more technical detail. But I'm making assumptions that documentation would be published I could link to. I'm happy to re-look at it in the modal itself if necessary?

marshallmain commented on x-pack/plugins/security_solution/server/lib/risk_engine/schema/risk_score_apis.yml:
```diff
@@ -44,6 +44,56 @@ paths:
                 $ref: '#/components/schemas/RiskScoresPreviewResponse'
         '400':
           description: Invalid request
+  /engine/status:
```
Nice to see the new APIs defined in OpenAPI already! At some point, either in this PR or a follow up, could you or @rylnd move these OpenAPI specs to `/common/api/risk_engine/<route name>` similar to the other API domains? This way the code generation script will be able to find these schemas and we'll have all our OpenAPI schemas in one folder (`/common/api`).

nkhristinin commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into enable-risk-score

nkhristinin committed:
Remove last updated_by

nkhristinin committed:
fix hest integration tests

nkhristinin committed:
Fix cypress tetss

nkhristinin committed:
Fix jest tests

MadameSheema commented on x-pack/plugins/security_solution/cypress/e2e/entity_analytics/entity_analytics_management_page.cy.ts:
```diff
@@ -14,81 +14,144 @@ import {
+      });
+
+      it('show error panel if API returns error and then try to refetch data', () => {
+        cy.intercept('POST', '/internal/risk_score/preview', {
```
NIT: Wrapt these intercepts into a task method making clear what you are doing or just add a simple comment about the stubbing you are doing :)

MadameSheema commented on x-pack/plugins/security_solution/cypress/e2e/entity_analytics/entity_analytics_management_page.cy.ts:
```diff
@@ -14,81 +14,144 @@ import {
+          },
+        });
+
+        cy.get(RISK_PREVIEW_ERROR_BUTTON).click();
```
NIT: Wrap this into a task method.

MadameSheema commented on x-pack/plugins/security_solution/cypress/e2e/entity_analytics/entity_analytics_management_page.cy.ts:
```diff
@@ -14,81 +14,144 @@ import {
-        body: {
-          scores: { host: [], user: [] },
-        },
+        cy.intercept('POST', '/internal/risk_score/engine/init', {
```
Same for this intercept.

MadameSheema commented on x-pack/plugins/security_solution/cypress/tasks/entity_analytics.ts:
```diff
@@ -31,3 +36,16 @@ export const enableJob = () => {
+};
+
+export const updateRiskEngineConfirm = () => {
+  cy.get(RISK_SCORE_UPDATE_CONFIRM).click();
```
Why this action is wrapped in `updateRiskEngineConfirm` and not part of `updateRiskEngine`?

nkhristinin replied:
It open modal, and there some ways to cancel this model, so I want to split those actions

MadameSheema reviewed (approved):
sec-eng-prod changes looked good to me!

Just left some nit comments. Thanks for increasing the test coverage Nikita!! :)

machadoum reviewed (commented):
threat-hunting-explore changes LGTM!

You need to check for `entity-analytics` UI privileges to hide the new page and the update panel for serverless essentials.

<img width="1783" alt="Screenshot 2023-08-03 at 12 07 59" src="https://github.com/elastic/kibana/assets/1490444/68a4e260-4a83-4e99-8aec-68ab261d1876">
<img width="1773" alt="Screenshot 2023-08-03 at 12 08 08" src="https://github.com/elastic/kibana/assets/1490444/56975b6d-f96d-4b29-ab45-c5f3c31162ab">

nkhristinin committed:
fix cypress tests

nkhristinin committed:
Access for license and serverless

nkhristinin committed:
Chaange approach for link with license and capabilities

nkhristinin committed:
Fix links

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

nkhristinin committed:
fix ts problems

dmlemeshko reviewed (approved):
Appex QA changes LGTM

tomsonpl commented on x-pack/plugins/security_solution/public/management/links.ts:
```diff
@@ -218,7 +218,8 @@ export const getManagementFilteredLinks = async (
       : getEndpointAuthzInitialState();
-
+  const isPlatinumOrTrialLicense =
+    licenseService.isPlatinumPlus() || licenseService.getLicenseInformation()?.type === 'trial';
```
Is there any specific reason to check for TRIAL? Isn't trial above Platinum? ;)

rylnd replied:
Agreed, addressed in 1bc331426d153b39beebec5b15063082795e2247.

tomsonpl reviewed (commented):
Nice job :+1: 
I left just one question as a Defend Workflows review :)

rylnd committed:
Merge branch 'main' into enable-risk-score

rylnd committed:
Simplify logic for showing EA management page

Trial licenses are included in `.isPlatinumPlus`, we don't need to check
that ourselves.

tomsonpl reviewed (approved):
Defend Workflows lgtm 👍

machadoum commented:
> threat-hunting-explore changes LGTM!
> 
> You need to check for `entity-analytics` UI privileges to hide the new page and the update panel for serverless essentials.
> 
> <img alt="Screenshot 2023-08-03 at 12 07 59" width="1783" src="https://user-images.githubusercontent.com/1490444/258079065-68a4e260-4a83-4e99-8aec-68ab261d1876.png"> 
Hey!
I took a second look and the call to action on the entity analytics page is still showing for serverless essentials.

nkhristinin committed:
Fix tesxt

nkhristinin committed:
Hide update panel for serverless

nkhristinin commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into enable-risk-score

nkhristinin commented:
@elasticmachine merge upstream

kibanamachine commented:
ignoring request to update branch, pull request is closed
